### DISCUSSION TITLE
SUBSTR performance

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#89] [Progress Updates]
my substr proposal (using BCT and conchar) runs in a 10 million loop around 6 secs.
@rvjansen : I am not sure, what the easiest way to call your substr 10 million times. 
```
main()   .locals=16
   load  r15,10000000
   time  r14
   itos r14
   say   r14
outer:
/*                   1         2         3         4
            1234567890123456789012345678901234567890123  */
   load r1,"The quick brown fox jumps over the lazy dog"
   load r8,"?"  /* padding byte */
   load r6,43   /* maxlen */
   load r7,0    /* over sized */
   load r2,""   /* extracted substr */
   load r4,21   /* substr offset */
   load r5,10    /* substr length */

substr:
   iadd r12,R4,R5   /* max offset length=offset+requested length */
   dec r12
   igte r11,r6,r12  /* requested length <= maxlength? */
   brt  lenok,r11   /* if true skip padding size calc */
   isub r7,r12,r6   /* calculate padding size */
   isub r5,r5,r7    /* reduce to maxlen */
 lenok:
   dec r4
loop:
   concchar r2,r1,r4
   bct loop,r5,r4
   ieq r11,r7,0     /* is requested length oversize? */
   brt nopad,r11    /* branch if not */
 pad:
   concat r2,r2,r8  /* we are past length, cat pad */
   itos r7
   say r7
   say r2
   bct    pad,r7
nopad:
   bct outer,r15
   time  r14
   itos r14
   say   r14
   say r2   /* output extracted substr */
   ret``

**Comment by rvjansen** [comment]
Hi Peter,

Just check it in - I am the egoless programmer. We can now test it with a Rexx script!

**Reply by adesutherland** [reply]
> Hi Peter, Just check it in - I am the egoless programmer. We can now test it with a Rexx script!

Interesting to compare pass by value & reference

    arg string1 = .string

with 

    arg string .string

In modern C++ it kind of passes by reference and duplicates on a write lazily. But anyway we may need a constant pass by reference option?

**Reply by Peter-Jacob** [reply]
@René : substr checked in as MID(string,offset,length,padding)

**Comment by adesutherland** [comment]
Can you explain bct - is the register reduced and compared to zero - i.e. loops r2 times?

REXX developer's loops generally count upwards - could we have a version of bct that counts upwards to the r3 limit or something like that? I think I could change the rxc to use that

**Reply by Peter-Jacob** [reply]
@adrian: you are right BCT label,count-down-var reduces the register by -1 every time the instruction is entered. To do it the other way round is theoretically possible, but then you need a max counter as an additional parameter. This will collide with the second version of 
BCT   label,count-down-var,go-up-var in which a second variable is additionally increased by +1, there is no space for an upper limit parm.
Concerning the rexx compiler it can't easily be used for do-loop, as BCT is executed at the end of a loop (the END statement), whereas the counter setting must be done at the beginning (the DO statement).

**Comment by Peter-Jacob** [comment]
@egoless programmer: stuff is checked in (note the label is now the first parameter) . I'd appreciate a new rxc version.

**Comment by rvjansen** [comment]
I saw mid() - looks great! As it is not directly comparable, I'll work the core of it into substr() and then will compare the instruction counts.

**Reply by rvjansen** [reply]
I opened an issue for a perceived problem I encountered doing that merge
